boututils/datafile.py

```diff
@@ -523,11 +541,15 @@ def write(self, name, data, info=False):
             # Not found, so add.
 
             # Get dimensions
-            defdims = [(),
-                       ('t',),
-                       ('x', 'y'),
-                       ('x', 'y', 'z'),
-                       ('t', 'x', 'y', 'z')]
+            try:
+                defdims = self._bout_dimensions_from_type(data.attributes['bout_type'])
+            except AttributeError:
+                defdims_list = [(),
+                                ('t',),
+                                ('x', 'y'),
+                                ('x', 'y', 'z'),
```

Comment:
Should `FieldPerp` be a part of this?
```suggestion
                                ('x', 'z'),
                                ('x', 'y', 'z'),
```

Reply:
Unfortunately, we can't include `FieldPerp` here, because we're detecting the dimensions of the array just from the dimensionality of the array, so couldn't tell the difference between `Field2D` and `FieldPerp`. A `FieldPerp` has to be passed as a `BoutArray` so that it has a `bout_type`.